### ISSUE TITLE
JEG 2.3 close websocket on large? stream data in spark_yarn_kernel

### ISSUE THREAD

**Issue by ying1** [jupyter/enterprise_gateway#914]
## Description

This is an issue observed w/ JEG 2.3 . This is not an issue that we have observed w/ JEG 2.1 . 

When connecting from elyra/2.0.0 notebook to JEG 2.3 to a `spark_yarn_kernel` , when issuing `!env` - the websocket disconnects.  The websocket is closed, while the kernel remains active.  

Issuing a Kernel -> Reconnect from the UI does resume the session. 

## Screenshots / Logs
NOTE: `!env` remains `*` - and did not return.  The cell post `!env` can be executed after reconnect is issued.
![image](https://user-images.githubusercontent.com/12278773/101873098-313b2100-3b3b-11eb-95b6-9247bc166b9e.png)

Logs on JEG side - showing websocket closed, while kernel is still active.
```
[D 2020-10-14 08:33:10.746 EnterpriseGatewayApp] activity on bb1fbd74-f170-4cef-86ca-c3c6ccd9cc77: status (busy)
[D 2020-10-14 08:33:10.748 EnterpriseGatewayApp] activity on bb1fbd74-f170-4cef-86ca-c3c6ccd9cc77: status (idle)
[D 2020-10-14 08:33:10.835 EnterpriseGatewayApp] activity on bb1fbd74-f170-4cef-86ca-c3c6ccd9cc77: status (busy)
[D 2020-10-14 08:33:10.837 EnterpriseGatewayApp] activity on bb1fbd74-f170-4cef-86ca-c3c6ccd9cc77: status (idle)
[D 2020-10-14 08:33:48.291 EnterpriseGatewayApp] Websocket closed bb1fbd74-f170-4cef-86ca-c3c6ccd9cc77:e045e623-72a3ff4c28b98fa479e509cc
[I 2020-10-14 08:33:48.291 EnterpriseGatewayApp] Starting buffering for bb1fbd74-f170-4cef-86ca-c3c6ccd9cc77:e045e623-72a3ff4c28b98fa479e509cc
[D 2020-10-14 08:33:48.291 EnterpriseGatewayApp] Clearing buffer for bb1fbd74-f170-4cef-86ca-c3c6ccd9cc77
[I 201014 08:34:00 web:2250] 200 GET /api/kernels/bb1fbd74-f170-4cef-86ca-c3c6ccd9cc77?doAs=user1 (127.0.0.1) 0.88ms
[I 201014 08:34:01 web:2250] 200 GET /api/kernels/bb1fbd74-f170-4cef-86ca-c3c6ccd9cc77?doAs=user1 (127.0.0.1) 0.79ms
```

## Environment
- Enterprise Gateway Version 2.3
- Platform: YARN / with Knox
- Others: elyra/2.0.0 docker container as notebook client. 

Launch script:
```
docker run -t --rm \
--name localnb2kg \
-p 8888:8888 \
-e KG_URL="https://${gateway}/jkg" \
-e KG_WS_URL="wss://${gateway}/jkgws"  \
-e KERNEL_USERNAME=user1 \
-e VALIDATE_KG_CERT=no \
-e KG_HEADERS='{"Authorization": "Bearer '${authToken}'","Content-Type": "application/json"}' \
-e KERNEL_EXTRA_SPARK_OPTS='--archives hdfs:///user/dsxhi/environments/pythonAddons/jeg_base.tar.gz --conf spark.yarn.appMasterEnv.PYSPARK_PYTHON=jeg_base.tar.gz/jeg_base/bin/python --conf spark.driver.memory=2g  --conf spark.driver.cores=1  --conf spark.executor.memory=1g --conf spark.executor.instances=2' \
--entrypoint="" \
elyra/nb2kg:2.0.0  /opt/conda/bin/jupyter notebook --NotebookApp.token='' \
--NotebookApp.session_manager_class=nb2kg.managers.SessionManager \
--NotebookApp.kernel_manager_class=nb2kg.managers.RemoteKernelManager \
--NotebookApp.kernel_spec_manager_class=nb2kg.managers.RemoteKernelSpecManager \
--NotebookApp.ip=0.0.0.0
```

**Comment by kevin-bates** [comment]
Could you please provide the expected output of the `!env` command so an applicable reproduction can take place?  This is not reproducible using "standard" environments.

Also, could you please reproduce the issue w/o Knox configured?  Are there any data-throttling limits configured in the reverse proxy?

If you ran `!ls -lR` against a reasonably high-level directory, does the problem reproduce there as well?  That command might provide a way to help determine any kind of limit that might be getting exceeded.

Thanks.

**Comment by ying1** [comment]
Hi @kevin-bates 
Sorry for the delay.  I've included the environment content below.   This is generated via:
```
!env > out.env.txt
```
which then has the host information removed
[out.env.redacted.txt](https://github.com/jupyter/enterprise_gateway/files/5784129/out.env.redacted.txt)

The websocket itself will also terminate if I tried to do the following. 
```
!cat out.env.redacted.txt
```
I'll check to see what happens without knox configured.

**Comment by kevin-bates** [comment]
Thanks @ying1 - I think that's a good test.  You might try to see if there are any data limits configured in your Knox settings.

FWIW, I copied `out.env.redacted.txt` to the YARN node in which I had a spark-yarn kernel running and could `cat` the file.  It returned with output almost immediately (DEBUG entries follow the screenshot)...

![image](https://user-images.githubusercontent.com/22599560/103951061-dc1b0c80-50f2-11eb-854c-7416aae89dc3.png)

```
[D 2021-01-07 14:12:53.102 EnterpriseGatewayApp] activity on fae36c3b-89fa-4c9c-bba8-4767eef1ebe5: status (busy)
[D 2021-01-07 14:12:53.103 EnterpriseGatewayApp] activity on fae36c3b-89fa-4c9c-bba8-4767eef1ebe5: execute_input
[D 2021-01-07 14:12:53.149 EnterpriseGatewayApp] activity on fae36c3b-89fa-4c9c-bba8-4767eef1ebe5: stream
[D 2021-01-07 14:12:53.258 EnterpriseGatewayApp] activity on fae36c3b-89fa-4c9c-bba8-4767eef1ebe5: status (idle)
```

**Comment by ying1** [comment]
@kevin-bates Thanks.  Will give that a try

**Comment by ying1** [comment]
@kevin-bates 
Looks like it is caused by Knox.  I am able to run the same code path `!env`, as well as using the same file to do `!cat out.env.redacted.txt`  without issues.

**Comment by kevin-bates** [comment]
Hi @ying1 - thank you for the update.  Hopefully, there's a knob that can be adjusted in Knox's configuration to address this.  If you find such a setting, could you please post the resolution here?

Thanks.

**Comment by ying1** [comment]
@kevin-bates 
Thanks.  Found a few defect w/ KNOX relating to websockets.  And indee there was an error logged in gateway.log that we did not observe. 
```
gateway.websockets (ProxyWebSocketAdapter.java:cleanupOnError(207)) - Error: org.eclipse.jetty.websocket.api.MessageTooLargeException: Text message size [37498] exceeds maximum size [32768]
```

- https://issues.apache.org/jira/browse/KNOX-836 - Indicates issue with Zepplin (uses websocket)
- https://issues.apache.org/jira/browse/KNOX-1828 - Indicates issue with Knox not using these configuration.

After adding the sample configuration from KNOX-1828 , I am able to pull the `!env` properly.
